File: apps/domain-hub/app/domain/slot_service.py
import datetime
import uuid
from typing import List, Optional, Set
import pytz
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_
from app.core.logging import get_logger
from app.db.models import User, EventType, AvailabilityRule, AvailabilityException, Slot, Booking
from app.domain.slot_generation import (
    TimeWindow,
    apply_exceptions_for_date,
    merge_windows,
    overlaps_booked,
    split_into_slots,
    windows_for_weekday_rule,
    parse_iso_date,
)

logger = get_logger("SlotService")

def generate_slot_id() -> str:
    return f"c{uuid.uuid4().hex[:24]}"

def regenerate_host_slots(
    db: Session,
    host_id: int,
    from_date_str: Optional[str] = None,
    to_date_str: Optional[str] = None,
    days_ahead: int = 30,
) -> int:
    host = db.query(User).filter(User.id == host_id).first()
    if not host:
        logger.warning(f"Host {host_id} not found for slot regeneration")
        return 0

    now_utc = datetime.datetime.now(pytz.UTC)
    today = now_utc.date()

    if from_date_str:
        from_date = parse_iso_date(from_date_str)
    else:
        from_date = today

    if to_date_str:
        to_date = parse_iso_date(to_date_str)
    else:
        to_date = from_date + datetime.timedelta(days=days_ahead)

    logger.info(
        f"Regenerating slots for host {host.name} (id={host_id}) from {from_date} to {to_date}",
        extra={"host_id": host_id, "from_date": str(from_date), "to_date": str(to_date)},
    )

    from_dt = datetime.datetime(from_date.year, from_date.month, from_date.day, 0, 0, 0, tzinfo=pytz.UTC)
    to_dt = datetime.datetime(to_date.year, to_date.month, to_date.day, 23, 59, 59, tzinfo=pytz.UTC)

    # 1. Fetch active rules
    rules = db.query(AvailabilityRule).filter(
        AvailabilityRule.userId == host_id,
        AvailabilityRule.isActive == True
    ).all()

    # 2. Fetch exceptions in range
    exceptions = db.query(AvailabilityException).filter(
        AvailabilityException.userId == host_id,
        AvailabilityException.date >= from_date,
        AvailabilityException.date <= to_date
    ).all()

    # 3. Fetch active event types for host
    event_types = db.query(EventType).filter(
        EventType.hostId == host_id,
        EventType.isActive == True
    ).all()

    if not event_types:
        logger.info(f"No active event types for host {host_id}, skipping slot generation")
        return 0

    # 4. Fetch existing non-canceled bookings in range
    bookings = db.query(Booking).filter(
        Booking.hostId == host_id,
        Booking.status != "CANCELED",
        Booking.startAt <= to_dt,
        Booking.endAt >= from_dt
    ).all()

    booked_windows = [
        TimeWindow(
            start=b.startAt if b.startAt.tzinfo else pytz.UTC.localize(b.startAt),
            end=b.endAt if b.endAt.tzinfo else pytz.UTC.localize(b.endAt),
        )
        for b in bookings
    ]

    total_created_or_updated = 0

    # 5. Generate slots per event type
    for event_type in event_types:
        generated_slot_keys: Set[str] = set()

        curr = from_date
        while curr <= to_date:
            weekday = (curr.weekday() + 1) % 7 # Luxon weekday: 0=Sun, 1=Mon, ..., 6=Sat

            # Build available windows for this date from rules
            day_windows: List[TimeWindow] = []
            for rule in rules:
                if rule.weekday == weekday:
                    day_windows.extend(windows_for_weekday_rule(rule, curr, host.timezone or "UTC"))

            # Merge overlapping windows
            day_windows = merge_windows(day_windows)

            # Apply date exceptions
            day_windows = apply_exceptions_for_date(day_windows, exceptions, curr, host.timezone or "UTC")

            # Split into discrete event-duration slots
            slots = split_into_slots(
                windows=day_windows,
                duration_minutes=event_type.durationMinutes,
                buffer_before=event_type.bufferBeforeMinutes,
                buffer_after=event_type.bufferAfterMinutes,
            )

            # Filter past slots and booked collisions
            for slot_win in slots:
                if slot_win.start < now_utc:
                    continue
                if overlaps_booked(slot_win, booked_windows):
                    continue

                start_iso = slot_win.start
                end_iso = slot_win.end
                slot_key = f"{event_type.id}|{start_iso.isoformat()}|{end_iso.isoformat()}"
                generated_slot_keys.add(slot_key)

                # Upsert slot into DB
                existing_slot = db.query(Slot).filter(
                    Slot.eventTypeId == event_type.id,
                    Slot.startAt == start_iso,
                    Slot.endAt == end_iso,
                ).first()

                if existing_slot:
                    if existing_slot.status != "BOOKED":
                        existing_slot.status = "AVAILABLE"
                        existing_slot.updatedAt = datetime.datetime.utcnow()
                else:
                    new_slot = Slot(
                        id=generate_slot_id(),
                        hostId=host_id,
                        eventTypeId=event_type.id,
                        startAt=start_iso,
                        endAt=end_iso,
                        status="AVAILABLE",
                        createdAt=datetime.datetime.utcnow(),
                        updatedAt=datetime.datetime.utcnow(),
                    )
                    db.add(new_slot)
                    total_created_or_updated += 1

            curr += datetime.timedelta(days=1)

        # Mark obsolete slots as BLOCKED
        existing_future_slots = db.query(Slot).filter(
            Slot.eventTypeId == event_type.id,
            Slot.startAt >= from_dt,
            Slot.startAt <= to_dt,
            Slot.status.in_(["AVAILABLE", "BLOCKED"])
        ).all()

        for slot in existing_future_slots:
            slot_start = slot.startAt if slot.startAt.tzinfo else pytz.UTC.localize(slot.startAt)
            slot_end = slot.endAt if slot.endAt.tzinfo else pytz.UTC.localize(slot.endAt)
            key = f"{event_type.id}|{slot_start.isoformat()}|{slot_end.isoformat()}"
            if key not in generated_slot_keys:
                slot.status = "BLOCKED"
                slot.updatedAt = datetime.datetime.utcnow()

    db.commit()
    logger.info(
        f"Slot regeneration finished for host {host_id}. New/updated slots: {total_created_or_updated}",
        extra={"host_id": host_id, "generated_count": total_created_or_updated},
    )
    return total_created_or_updated

File: hubs/domain-hub/src/services/slot.service.ts
import { DateTime } from 'luxon'
import { slotRepository } from '../repositories/slot.repository.js'
import { userRepository } from '../repositories/user.repository.js'
import { eventTypeRepository } from '../repositories/event-type.repository.js'
import { availabilityRepository } from '../repositories/availability.repository.js'
import { bookingRepository } from '../repositories/booking.repository.js'
import {
  mergeWindows,
  subtractWindows,
  splitIntoSlots,
  overlapsBooked,
  parseTimeToDateTime,
  type TimeWindow,
} from './slot-generation.service.js'
import { ApiError } from '../utils/errors.js'
import { logger } from '../utils/logger.js'

export class SlotService {
  async getAvailableSlots(eventTypeId: number, dateStr: string) {
    const et = await eventTypeRepository.findById(eventTypeId)
    if (!et) {
      throw ApiError.notFound(`EventType ${eventTypeId} not found`)
    }

    const hostTz = et.host?.timezone || 'UTC'
    const startOfDay = DateTime.fromISO(dateStr, { zone: hostTz }).startOf('day').toUTC().toJSDate()
    const endOfDay = DateTime.fromISO(dateStr, { zone: hostTz }).endOf('day').toUTC().toJSDate()

    const slots = await slotRepository.findAvailableSlotsByDate(eventTypeId, startOfDay, endOfDay)
    return slots.map((s) => ({
      id: s.id,
      hostId: s.hostId,
      eventTypeId: s.eventTypeId,
      startAt: s.startAt.toISOString(),
      endAt: s.endAt.toISOString(),
      status: s.status,
      createdAt: s.createdAt.toISOString(),
      updatedAt: s.updatedAt.toISOString(),
      eventType: s.eventType
        ? {
            id: s.eventType.id,
            hostId: s.eventType.hostId,
            title: s.eventType.title,
            slug: s.eventType.slug,
            durationMinutes: s.eventType.durationMinutes,
            isActive: s.eventType.isActive,
            locationType: s.eventType.locationType,
            locationValue: s.eventType.locationValue,
            bufferBeforeMinutes: s.eventType.bufferBeforeMinutes,
            bufferAfterMinutes: s.eventType.bufferAfterMinutes,
            createdAt: s.eventType.createdAt.toISOString(),
            updatedAt: s.eventType.updatedAt.toISOString(),
          }
        : null,
    }))
  }

  async listSlotsForHost(hostId: number, from?: string, to?: string) {
    const fromDate = from ? new Date(from) : undefined
    const toDate = to ? new Date(to) : undefined
    const slots = await slotRepository.findSlotsForHost(hostId, fromDate, toDate)
    return slots.map((s) => ({
      id: s.id,
      hostId: s.hostId,
      eventTypeId: s.eventTypeId,
      startAt: s.startAt.toISOString(),
      endAt: s.endAt.toISOString(),
      status: s.status,
      createdAt: s.createdAt.toISOString(),
      updatedAt: s.updatedAt.toISOString(),
      eventType: s.eventType
        ? {
            id: s.eventType.id,
            hostId: s.eventType.hostId,
            title: s.eventType.title,
            slug: s.eventType.slug,
            durationMinutes: s.eventType.durationMinutes,
            isActive: s.eventType.isActive,
            locationType: s.eventType.locationType,
            locationValue: s.eventType.locationValue,
            bufferBeforeMinutes: s.eventType.bufferBeforeMinutes,
            bufferAfterMinutes: s.eventType.bufferAfterMinutes,
            createdAt: s.eventType.createdAt.toISOString(),
            updatedAt: s.eventType.updatedAt.toISOString(),
          }
        : null,
    }))
  }

  async createSlot(data: { hostId: number; eventTypeId: number; startAt: string; endAt: string }) {
    const slotId = `c${Date.now().toString(36)}${Math.random().toString(36).substring(2, 8)}`
    const slot = await slotRepository.upsert({
      id: slotId,
      hostId: data.hostId,
      eventTypeId: data.eventTypeId,
      startAt: new Date(data.startAt),
      endAt: new Date(data.endAt),
      status: 'AVAILABLE',
    })
    return {
      id: slot.id,
      hostId: slot.hostId,
      eventTypeId: slot.eventTypeId,
      startAt: slot.startAt.toISOString(),
      endAt: slot.endAt.toISOString(),
      status: slot.status,
      createdAt: slot.createdAt.toISOString(),
      updatedAt: slot.updatedAt.toISOString(),
    }
  }

  async updateSlot(
    id: string,
    hostId: number,
    data: { eventTypeId?: number; startAt?: string; endAt?: string; status?: string }
  ) {
    const existing = await slotRepository.findById(id)
    if (!existing || existing.hostId !== hostId) {
      throw ApiError.forbidden(`Cannot update slot ${id}`)
    }
    const updated = await slotRepository.upsert({
      id,
      hostId,
      eventTypeId: data.eventTypeId || existing.eventTypeId,
      startAt: data.startAt ? new Date(data.startAt) : existing.startAt,
      endAt: data.endAt ? new Date(data.endAt) : existing.endAt,
      status: data.status || existing.status,
    })
    return {
      id: updated.id,
      hostId: updated.hostId,
      eventTypeId: updated.eventTypeId,
      startAt: updated.startAt.toISOString(),
      endAt: updated.endAt.toISOString(),
      status: updated.status,
      createdAt: updated.createdAt.toISOString(),
      updatedAt: updated.updatedAt.toISOString(),
    }
  }

  async deleteSlot(id: string, hostId: number) {
    const existing = await slotRepository.findById(id)
    if (!existing || existing.hostId !== hostId) {
      throw ApiError.forbidden(`Cannot delete slot ${id}`)
    }
    await slotRepository.delete(id)
    return { success: true, message: `Slot ${id} deleted` }
  }

  async regenerateSlots(hostId: number, fromDateStr?: string, toDateStr?: string, daysAhead = 30) {
    const user = await userRepository.findById(hostId)
    if (!user) {
      throw ApiError.notFound(`User ${hostId} not found`)
    }

    const hostTz = user.timezone || 'UTC'
    const nowHost = DateTime.now().setZone(hostTz)

    const fromDate = fromDateStr
      ? DateTime.fromISO(fromDateStr, { zone: hostTz }).startOf('day')
      : nowHost.startOf('day')
    const toDate = toDateStr
      ? DateTime.fromISO(toDateStr, { zone: hostTz }).endOf('day')
      : fromDate.plus({ days: daysAhead }).endOf('day')

    logger.info(`Regenerating slots for host ${user.name} (${hostId}) from ${fromDate.toISODate()} to ${toDate.toISODate()}`)

    // 1. Fetch rules & exceptions & event types
    const rules = await availabilityRepository.findRulesByUserId(hostId)
    const activeRules = rules.filter((r) => r.isActive)

    const exceptions = await availabilityRepository.findExceptionsByUserId(
      hostId,
      fromDate.toJSDate(),
      toDate.toJSDate()
    )

    const eventTypes = await eventTypeRepository.findByHostId(hostId)
    const activeEventTypes = eventTypes.filter((e) => e.isActive)

    if (activeEventTypes.length === 0) {
      return { success: true, generatedCount: 0, message: 'No active event types for host' }
    }

    // 2. Fetch existing active bookings in range
    const activeBookings = await bookingRepository.findActiveBookingsInRange(
      hostId,
      fromDate.toUTC().toJSDate(),
      toDate.toUTC().toJSDate()
    )

    const bookedWindows: TimeWindow[] = activeBookings.map((b) => ({
      start: DateTime.fromJSDate(b.slot.startAt).toUTC(),
      end: DateTime.fromJSDate(b.slot.endAt).toUTC(),
    }))

    let generatedCount = 0

    // Iterate through days
    let dayCursor = fromDate
    while (dayCursor <= toDate) {
      const dateStr = dayCursor.toISODate()
      if (!dateStr) {
        dayCursor = dayCursor.plus({ days: 1 })
        continue
      }

      const weekday = dayCursor.weekday % 7 // 0=Sunday, 1=Monday... 6=Saturday
      const dayRules = activeRules.filter((r) => r.weekday === weekday)

      // Base windows for weekday
      let dayWindows: TimeWindow[] = []
      for (const rule of dayRules) {
        const start = parseTimeToDateTime(dateStr, rule.startTime, rule.timezone || hostTz)
        const end = parseTimeToDateTime(dateStr, rule.endTime, rule.timezone || hostTz)
        if (end > start) {
          dayWindows.push({ start, end })
        }
      }

      dayWindows = mergeWindows(dayWindows)

      // Apply exceptions for this date
      const dayExceptions = exceptions.filter(
        (ex) => DateTime.fromJSDate(ex.date).toISODate() === dateStr
      )

      for (const ex of dayExceptions) {
        if (ex.type === 'BLOCK_FULL_DAY') {
          dayWindows = []
        } else if (ex.type === 'BLOCK_PARTIAL' && ex.startTime && ex.endTime) {
          const blockStart = parseTimeToDateTime(dateStr, ex.startTime, ex.timezone || hostTz)
          const blockEnd = parseTimeToDateTime(dateStr, ex.endTime, ex.timezone || hostTz)
          dayWindows = subtractWindows(dayWindows, { start: blockStart, end: blockEnd })
        } else if (ex.type === 'ADD_AVAILABLE_WINDOW' && ex.startTime && ex.endTime) {
          const addStart = parseTimeToDateTime(dateStr, ex.startTime, ex.timezone || hostTz)
          const addEnd = parseTimeToDateTime(dateStr, ex.endTime, ex.timezone || hostTz)
          if (addEnd > addStart) {
            dayWindows.push({ start: addStart, end: addEnd })
            dayWindows = mergeWindows(dayWindows)
          }
        }
      }

      // Generate slots per event type
      for (const et of activeEventTypes) {
        const rawSlots = splitIntoSlots(
          dayWindows,
          et.durationMinutes,
          et.bufferBeforeMinutes,
          et.bufferAfterMinutes
        )

        for (const slot of rawSlots) {
          // Check if slot overlaps non-cancelled booking
          const isOverlapping = overlapsBooked(
            slot,
            bookedWindows,
            et.bufferBeforeMinutes,
            et.bufferAfterMinutes
          )

          if (!isOverlapping) {
            const slotId = `c${slot.start.toMillis().toString(36)}${Math.random().toString(36).substring(2, 6)}`
            await slotRepository.upsert({
              id: slotId,
              hostId,
              eventTypeId: et.id,
              startAt: slot.start.toJSDate(),
              endAt: slot.end.toJSDate(),
              status: 'AVAILABLE',
            })
            generatedCount++
          }
        }
      }

      dayCursor = dayCursor.plus({ days: 1 })
    }

    return {
      success: true,
      generatedCount,
      message: `Generated ${generatedCount} available slots for host ${user.name}`,
    }
  }
}

export const slotService = new SlotService()
